posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+
+val usersW = Witness("users")
+```
+
+But this isn't a pure type declaration, you will say. And you are right, but right now there is no other way in Scala. We have to go the ordinary value road first to create our types.
+
+Now that we know how to get a type representation from a `String` which describes our path we should clearly mark it as a path element:
+
+```scala
+sealed trait Path[P]
+
+type users = Path[usersW.T]
+```
+
+That's it. That is the basic concept of how we can describe our apis as types. We just reuse this concept now for the remaining elements like the segment.
+
+
+```scala
+val nameW = Witness('name)
+
+sealed trait Segment[K, V]
+
+type name = Segment[nameW.T, String]
+```
+
+Do you see how we included the segment's identifier in the type? This way we are not only gain information about the expected type but also what kind of value we want to see. By the way, I decided to use `Symbols` as identifiers, but you could also switch to `String` literals. The remaining definitions look pretty similar:
+
+```scala
+val minAgeW = Witness('minAge)
+
+sealed trait Query[K, V]
+
+type minAge = Query[minAgeW.T, Int]
+
+sealed trait Method
+sealed trait Get[A] extends Method
+```
+
+Now that we know how to obtain the types of our api elements we have to put them together into a single type representation. After looking through shapeless's features we will find `HLists`, a list structure which can store elements of different types.
+
+```scala
+import shapeless.{::, HNil}
+
+type Api = Get[List[User]] :: users :: name :: minAge :: HNil
+```
+
+Here you go. `Api` is an exact representation of the endpoint we defined at the beginning. But you don't want to write `Witness` and `HLists` all the time so let's wrap it up into a convenient function call:
+
+```scala
+def api[M <: Method, P <: HList, Q <: HList, Api <: HList]
+       (method: M, path: PathList[P], queries: QueryList[Q])
+       (implicit prepQP: Prepend.Aux[Q, P, Api]): ApiTypeCarrier[M :: Api] = ApiTypeCarrier()
+      
+val Api = api(Get[List[User]], Root / "users" / Segment[String]('name), Queries.add(Query[Int]('minAge)))
+```
+
+Not clear what is happening? Let's take a look at the different elements of `def api(...)`:
+  * `method` should be obvious. It takes some method type.
+  * `PathList` is a type carrier with a function `def /(...)` to concatenate path elements and segments. In the end, `PathList` only stores the type of an `HList` and nothing more.
+  
+```scala
+final case class PathList[P <: HList]() {
+  
+  def /[S](path: Witness.Lt[S]): PathList[S :: P] = PathList()
+  ...
+}
+
+val Root = PathList[HNil]()
+```
+  * Same is true for `QueryList`.
+  * The last step is to merge all these `HLists` types into a single one. Shapeless comes again with a handy type class called `Prepend` which provides us with the necessary functionality. Two `HList` types go in, a single type comes out. And again, we use a type carrier here to store the api type.
```

Comment:
I prefer to think about typeclasses (when used for typelevel programming like here) as constraints rather than functions. So I'd say that `Prepend.Aux[Q, P, Api]` makes sure (or constrains) that `Api` is indeed the concatenation of `P` and `Q`. But that's a personal POV.

Reply:
My 2 Cents: it depends on what I try to achieve. If I want to proof something like `implicit ev: Monad[F]` it constraints the (a at some point known) type. But when computing something on the type level (like the concatenation) I would tend to describe it as function (computation), because here `Api` is solely determined by this computation.

But yes, in the end you could also describe it as a constraint put on `Api` :).